src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestDiscoverer.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Features.Testing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Testing;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.TestPlatform.VsTestConsole.TranslationLayer;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Testing;
+
+[Export, Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal partial class TestDiscoverer(ILoggerFactory loggerFactory)
+{
+    /// <summary>
+    /// TODO - localize messages. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1799066/
+    /// </summary>
+    private const string StageName = "Discovering tests...";
+    private readonly ILogger _logger = loggerFactory.CreateLogger<TestDiscoverer>();
+
+    /// <summary>
+    /// Finds tests in the specified document in the specified range.
+    ///
+    /// Note that since tests run against the last built dll,
+    /// </summary>
+    public async Task<ImmutableArray<TestCase>> DiscoverTestsAsync(
+        LSP.Range range,
+        Document document,
+        string projectOutputPath,
+        BufferedProgress<RunTestsPartialResult> progress,
+        VsTestConsoleWrapper vsTestConsoleWrapper,
+        CancellationToken cancellationToken)
+    {
+        var partialResult = new RunTestsPartialResult(StageName, $"{Environment.NewLine}Starting test discovery", Progress: null);
+        progress.Report(partialResult);
+
+        var testMethodFinder = document.GetRequiredLanguageService<ITestMethodFinder>();
+
+        // Find any potential test methods (based on attributes) that exist in the input range.
+        var potentialTestMethods = await GetPotentialTestMethodsAsync(range, document, testMethodFinder, cancellationToken);
```

Comment:
If I try to run tests in class B, and the tests are defined in the base class, will this logic work?
```
class B : A  { 

}
class A {
    [Fact]
    public void Test() { }
}
```

Reply:
This also won't work since we look only at the attribute syntax on methods.  But this also doesn't work in O# either.

Reply:
i'm fine with this not working.